docs/way-of-working/guides/testing/testing.md

```diff
@@ -1 +1,132 @@
-## Testing overview
\ No newline at end of file
+---
+title: Way of Testing
+sidebar_position: 10
+---
+
+:::danger[Draft]
+
+This document is still in draft. You can discuss it as part of this PR: https://github.com/dhis2/wow-frontend/pull/7.
+
+The scope of this document is automated tests _in code_; unit, integration, end to end, contract etc... in the context of the _frontend_ for platform team. It does not tackle other important aspects such as manual testing and its processes, or backend and database tests.
+:::
+
+### Why automated tests?
+
+Why we write tests? Good tests can...
+
+1. Verify the code is working correctly...
+2. Prevent future regressions
+3. Document code behaviour
+4. Provides design guidance
+5. Support refactoring
+
+> Read the full article here:  https://madeintandem.com/blog/five-factor-testing/
+
+
+:::warning[Discussion]
+
+- Rank these reasons in terms of importance in our context according to you
+- Any other reasons we should write tests?
+
+:::
+
+
+### The ideal scenario
+
+(Mozafar): we have a **high coverage** of tests, _integration-like_ with the most crucial scenarios tested in a thin layer of _end to end_ tests. Regressions and breaking changes in the API are caught early by tests, and refactoring made easy.
+
+:::warning[Discussion]
+
+- How does the ideal scenario look for you? some other possible prompts: TDD, cucumber, contract testing, end to end, pair programming... 
```

Comment:
I think that the term "integration test" is quite ambiguous and has a different scope from developer to developer.
Although I agree that integration tests (the way I understand it) should end up being the majority of tests in an hollistic test suite, stating that we should write mostly "integration-like" tests is not a guideline for what we actually want to test.

That's why I would refrain from using that term in the first place and change the perspective: Instead of testing abstractions, the code that uses abstractions should be tested.

The idea behind this approach is:
* it's much less ambiguous what test to code
* It allows us to change abstractions without having to change tests (the code that uses the abstractions should still work)
* It allows us to change a single "callee" of an abstraction and adjust its test code without having to touch anything related to the abstraction. If we realize during this that the abstraction is actually wrong for this particular case, we have a lot of options: 
  * We can change the abstraction and all "callees" will tell us whether we changes it correctly
  * We can inline the abstraction as we don't need an abstraction yet when the needed, new implementation of it is only relevant to a single "callee"
  * We can easily create a new abstraction for all "callees" that have used the wrong abstraction before if others still need the same abstraction without any changes. (Otherwise we'd have to add conditions to the abstraction, which obfuscates the original intention and therefore pollutes the implementation. We'd make the abstraction more complex as we'd mix together two different "ideas" of the abstraction in one place)

This [presentation](https://www.deconstructconf.com/2019/dan-abramov-the-wet-codebase) by Dan Abramov (creator of redux and facebook core developer for react) explains this in detail in an entertaining way.

Reply:
For e2e tests: The guideline I've used recently is: Cover critical user paths (from start to end) with details.